pkg/controllers/work/apply_controller.go

```diff
@@ -335,26 +339,38 @@ func (r *ApplyWorkReconciler) decodeManifest(manifest workv1alpha1.Manifest) (sc
 
 // Determines if an unstructured manifest object can & should be applied. If so, it applies (creates) the resource on the cluster.
 func (r *ApplyWorkReconciler) applyUnstructured(ctx context.Context, gvr schema.GroupVersionResource,
-	manifestObj *unstructured.Unstructured) (*unstructured.Unstructured, applyAction, error) {
+	manifestObj *unstructured.Unstructured, owner metav1.OwnerReference) (*unstructured.Unstructured, applyAction, error) {
 	manifestRef := klog.ObjectRef{
 		Name:      manifestObj.GetName(),
 		Namespace: manifestObj.GetNamespace(),
 	}
-	// compute the hash without taking into consider the last applied annotation
-	if err := setManifestHashAnnotation(manifestObj); err != nil {
+	manifestHash, err := computeManifestHash(manifestObj)
+	if err != nil {
 		return nil, ManifestNoChangeAction, err
 	}
 
 	// extract the common create procedure to reuse
 	var createFunc = func() (*unstructured.Unstructured, applyAction, error) {
-		// record the raw manifest with the hash annotation in the manifest
-		if err := setModifiedConfigurationAnnotation(manifestObj); err != nil {
+		// set config map name annotation on manifest
+		configMapName, err := r.getRandomConfigMapName(ctx)
+		if err != nil {
+			klog.ErrorS(err, "failed to get random config map name for manifest", "manifest", manifestObj.GetName(), "namespace", manifestObj.GetNamespace())
+			return nil, ManifestNoChangeAction, err
+		}
+		setConfigMapNameAnnotation(configMapName, manifestObj)
+		// record the raw manifest with config map name annotation
+		lastModifiedConfig, err := computeModifiedConfiguration(manifestObj)
+		if err != nil {
 			return nil, ManifestNoChangeAction, err
 		}
 		actual, err := r.spokeDynamicClient.Resource(gvr).Namespace(manifestObj.GetNamespace()).Create(
 			ctx, manifestObj, metav1.CreateOptions{FieldManager: workFieldManagerName})
 		if err == nil {
 			klog.V(2).InfoS("successfully created the manifest", "gvr", gvr, "manifest", manifestRef)
+			// create configmap
+			if err := r.createConfigMap(ctx, configMapName, owner, manifestHash, lastModifiedConfig); err != nil {
```

Comment:
Hi Arvind! The `createConfigMap` function is repeatedly called throughout the code; a little bit refactoring could help make the code more clear.

Reply:
Or more precisely, this place might not be the best one to do this create call.

Reply:
As we have discussed over the meeting, the name collision check currently in the PR might not work in adverse conditions (out-of-sync caches), and if the check fails to guard against name collisions, two manifests may share the same configmap and the behavior is undetermined.

Plus, w/o doubly links (or a guaranteed 1:1 name mapping), partial failures may also lead to undetermined behaviors.